store/store.go

```diff
@@ -245,82 +275,181 @@ func (bs *BlockStore) LoadSeenCommit(height int64) *types.Commit {
 	}
-	bs.mtx.RLock()
-	if height > bs.height {
-		bs.mtx.RUnlock()
-		return 0, fmt.Errorf("cannot prune beyond the latest height %v", bs.height)
+
+	if height > bs.Height() {
+		return 0, fmt.Errorf("height must be equal to or less than the latest height %d", bs.Height())
+	}
+
+	// remove block meta first as this is used to indicate whether the block exists.
+	// For this reason, we also use ony block meta as a measure of the amount of blocks pruned
+	pruned, err := bs.pruneBlockMetaAndHashes(height)
+	if err != nil {
+		return pruned, err
 	}
-	base := bs.base
-	bs.mtx.RUnlock()
-	if height < base {
-		return 0, fmt.Errorf("cannot prune to height %v, it is lower than base height %v",
-			height, base)
+
+	if err := bs.batchDelete(blockPartKey(1, 0), blockPartKey(height, 0)); err != nil {
+		return pruned, err
 	}
 
-	pruned := uint64(0)
+	if err := bs.batchDelete(blockCommitKey(1), blockCommitKey(height)); err != nil {
+		return pruned, err
+	}
+
+	if err := bs.batchDelete(seenCommitKey(1), seenCommitKey(height)); err != nil {
+		return pruned, err
+	}
+
+	return pruned, nil
+}
+
+// pruneBlockMetaAndHashes prunes all block metas from the base height to the retainHeight
+// but before each, loads the hash from the block meta to also prune from the block hash table
+func (bs *BlockStore) pruneBlockMetaAndHashes(retainHeight int64) (uint64, error) {
+	iter, err := bs.db.Iterator(
+		blockMetaKey(1),
+		blockMetaKey(retainHeight),
+	)
+	if err != nil {
+		panic(err)
+	}
+	defer iter.Close()
+
 	batch := bs.db.NewBatch()
 	defer batch.Close()
-	flush := func(batch dbm.Batch, base int64) error {
-		// We can't trust batches to be atomic, so update base first to make sure noone
-		// tries to access missing blocks.
-		bs.mtx.Lock()
-		bs.base = base
-		bs.mtx.Unlock()
-		bs.saveState()
-
-		err := batch.WriteSync()
-		if err != nil {
-			return fmt.Errorf("failed to prune up to height %v: %w", base, err)
-		}
-		batch.Close()
-		return nil
-	}
 
-	for h := base; h < height; h++ {
-		meta := bs.LoadBlockMeta(h)
-		if meta == nil { // assume already deleted
-			continue
-		}
-		if err := batch.Delete(blockMetaKey(h)); err != nil {
-			return 0, err
+	pruned := uint64(0)
+	for iter.Valid() {
+		// delete the block meta
+		if err := batch.Delete(iter.Key()); err != nil {
+			return 0, fmt.Errorf("pruning error at key %X: %w", iter.Key(), err)
 		}
-		if err := batch.Delete(blockHashKey(meta.BlockID.Hash)); err != nil {
-			return 0, err
+
+		// load the block meta so we can also remove the block hash from the db
+		var pbbm = new(tmproto.BlockMeta)
+		bz := iter.Value()
+
+		err = proto.Unmarshal(bz, pbbm)
+		if err != nil {
+			panic(fmt.Errorf("unmarshal to tmproto.BlockMeta: %w", err))
 		}
-		if err := batch.Delete(blockCommitKey(h)); err != nil {
-			return 0, err
+
+		blockMeta, err := types.BlockMetaFromProto(pbbm)
+		if err != nil {
+			panic(fmt.Errorf("error from proto blockMeta: %w", err))
```

Comment:
Why are we panicing when the function returns an error?

Reply:
good point. I was just copying behavior from another part of the code